Source/ImPlotWidgets/Private/HistogramWidget.cpp

```diff
@@ -5,9 +5,10 @@
 #include <ImGui/Public/ImPlotImplementation.h>
 #include <Runtime/Core/Public/Algo/Sort.h>
 
-void SHistogramWidget::Construct(const FArguments& InArgs)
+void SHistogramWidget::Construct(const FArguments& InArgs, TArray<float> InData)
```

Comment:
You should pass this TArray as a `const TArray&` Currently your code (nominally, as the compiler may be smart enough to optimise this away) is creating a TArray copy as part of the function call, then on line 11 you are copy assigning this array to your member variable. The result is two array constructions when you need one of those at most. There's another example of this pattern in SFilledGraphWidget::Construct.

Generally, unless it's a primitive data type (int, uintX, uint_X_t, float and so on) or occasionally a class / struct that has negligible construction overhead then you should pass params as const references.

Arrays and strings (which are specialised arrays) are expensive to construct as they have to make a dynamic (heap) allocation!

Reply:
Good spot @Josef-CL, would this have to be explicitly a const reference reference `const TArray<float>&&` to allow moving, or can you MoveTemp into a const reference fine? I can see lots of common scenarios where the data arrays would be constructed solely for use by the graph and so moving would be the best thing to do.

Reply:
Yeah if you want that to be the API then you would have to make it an R-value (&&) param toa loow move semantics. But not const as you explicitly want to take ownership and allow move to happen.